apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs

```diff
@@ -0,0 +1,150 @@
+            ApplyResourcePrefixHeader(ref settings, request.Name);
+
+        partial void Modify_ListSessionsRequest(ref ListSessionsRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_DeleteSessionRequest(ref DeleteSessionRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Name);
+
+        partial void Modify_ExecuteSqlRequest(ref ExecuteSqlRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_ExecuteBatchDmlRequest(ref ExecuteBatchDmlRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_ReadRequest(ref ReadRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_BeginTransactionRequest(ref BeginTransactionRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_CommitRequest(ref CommitRequest request, ref CallSettings settings)=>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_RollbackRequest(ref RollbackRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_PartitionQueryRequest(ref PartitionQueryRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_PartitionReadRequest(ref PartitionReadRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        private static void ApplyResourcePrefixHeader(ref CallSettings settings, string resource)
+        {
+            // If we haven't been given a resource name, just leave the request as it is.
+            if (string.IsNullOrEmpty(resource))
+            {
+                return;
+            }
+            string database = GetDatabaseResourceName(resource);
+            settings = settings.WithHeader(ResourcePrefixHeader, database);
+        }
+
+        // Visible for testing
+
+        /// <summary>
+        /// Retrieves the database resource name from a full resource name.
+        /// Validation is performed as far as the database ID but no further; the database ID is deemed to end at the first slash.
+        /// </summary>
+        /// <param name="resource">The resource name, which must start with projects/{project_id}/instances/{instance_id}/databases/{database_id}</param>
+        /// <returns>The database resource name</returns>
+        internal static string GetDatabaseResourceName(string resource)
```

Comment:
Drive-by as I don't have time to review *right* now - I strongly suspect this could be massively reduced using DatabaseName.TryParse or DatabaseName.Parse.

Reply:
Ah yes, that's a good idea. I will fix it early next week. Don't worry about reviewing until I fix it.

Reply:
That was much nicer. I still want to write more comprehensive tests for whether the header is set but at least logic-wise, this is hopefully now closer to the final implementation.